File: c++/LRUCache.cpp
//
//  LRUCache.cpp
//  c++
//
//  Created by sure on 2/1/15.
//  Copyright (c) 2015 Shuo Li. All rights reserved.
//

#include <stdio.h>
#include <unordered_map>

class Node {
public:
    Node *prev;
    Node *next;
    int val;
    int key;
    Node (int key, int val) {
        this->key = key;
        this->val = val;
        this->prev = NULL;
        this->next = NULL;
    }
};

class DoublyLinkedList {
public:
    DoublyLinkedList () {
        this->head = NULL;
        this->tail = NULL;
    }
    
    Node* head;
    Node* tail;
    
    void addFirst(Node* node) {
        if (!head) {
            node->prev = NULL;
            node->next = NULL;
            this->head = node;
            this->tail = node;
            return;
        }
        node->prev = NULL;
        this->head->prev = node;
        node->next =this->head;
        this->head = node;
    }
    
    void removeElement(Node* node) {
        if (this->head == this->tail) {
            this->head = NULL;
            this->tail = NULL;
            return;
        }
        if (this->head == node) {
            this->head = node->next;
            node->next->prev = NULL;
            return;
        }
        if (this->tail == node) {
            this->tail = node->prev;
            node->prev->next = NULL;
            return;
        }
        node->prev->next = node->next;
        node->next->prev = node->prev;
    }
};


class LRUCache{
public:
    int capacity;
    int size;
    std::unordered_map<int, Node*> dictionary;
    DoublyLinkedList *cache;
    
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
        this->cache = new DoublyLinkedList();
        this->size = 0;
    }
    
    int get(int key) {
        if (dictionary.find(key) == dictionary.end()){
            return -1;
        } else {
            this->cache->removeElement(dictionary[key]);
            this->cache->addFirst(dictionary[key]);
            return dictionary[key]->val;
        }
    }
    
    void set(int key, int value) {
        if (dictionary.find(key) != dictionary.end()) {
            dictionary[key]->val = value;
            this->cache->removeElement(dictionary[key]);
            this->cache->addFirst(dictionary[key]);
        } else {
            if (this->size < this->capacity) {
                this->size ++;
            } else {
                this->dictionary.erase(dictionary.find(this->cache->tail->key));
                this->cache->removeElement(this->cache->tail);
            }
            Node *node = new Node(key, value);
            this->dictionary[key] = node;
            this->cache->addFirst(dictionary[key]);
        }
    }
};

File: archive/python/LRUCache.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
    
    def removeNode(self, node):
        if self.head == self.tail:
            self.head = None
            self.tail = None
            return
        if self.head == node:
            self.head = node.next
            node.next.prev = None
            return
        if self.tail == node:
            self.tail = node.prev
            node.prev.next = None
            return
        node.prev.next = node.next
        node.next.prev = node.prev
        
    def addFirst(self, node):
        if not self.head:
            node.prev = None
            node.next = None
            self.head = node
            self.tail = node
            return
        
        node.head = None
        self.head.prev = node
        node.next = self.head
        self.head = node


class LRUCache:

    # @param capacity, an integer
    def __init__(self, capacity):
        self.D = {}
        self.cache = DoublyLinkedList()
        self.size = 0
        self.capacity = capacity

    # @return an integer
    def get(self, key):
        if key in self.D:
            self.cache.removeNode(self.D[key])
            self.cache.addFirst(self.D[key])
            return self.D[key].value
        else:
            return -1

    # @param key, an integer
    # @param value, an integer
    # @return nothing
    def set(self, key, value):
        if key in self.D:
            self.D[key].value = value
            self.cache.removeNode(self.D[key])
            self.cache.addFirst(self.D[key])
        else:
            if self.size < self.capacity:
                self.size += 1
            else:
                del self.D[self.cache.tail.key]
                self.cache.removeNode(self.cache.tail)
            node = Node(key, value)
            self.D[key] = node
            self.cache.addFirst(node)
